### PR TITLE
Enhanced Sort command

### PR DESCRIPTION
**Pull request by TanYikai**

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Person.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
     private ObjectProperty<UniqueTagList> tags;
     private ObjectProperty<UniqueRelationshipList> relationships;
 
+    private int sortOption;
```
Comment (Xenonym) [inline src/main/java/seedu/address/model/person/Person.java]: It seems really strange that each person would know how they would be sorted - what does it mean to have a single person have a sort order, especially when not in the context of the list?

File: src/main/java/seedu/address/model/person/Person.java [hunk 2]
```diff
@@ -235,9 +246,24 @@ public String toString() {
...
 
+    //@@author TanYikai
+    /**
+     * The compareTo method compares different attribute of the person object according to sortOption number
+     * The default is compare by name
+     * 1,2,3,4 represents phone, email, address amd remark respectively
+     */
     @Override
     public int compareTo(Person o) {
+        if (sortOption == 1) {
```
Comment (Xenonym) [inline src/main/java/seedu/address/model/person/Person.java]: What happens when `this` has a sort option that is different from `o`?

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 3]
```diff
@@ -72,10 +72,33 @@ public void setPerson(ReadOnlyPerson target, ReadOnlyPerson editedPerson)
     }
     //@@author TanYikai
     /**
-     * Sorts the persons object in the list alphanumerically by name.
+     * Sorts the persons object according to the sortOption.
+     * 0,1,2,3,4 represents name, phone, email, address, remark respectively
      */
-    public void sort() {
+    public void sort(int sortOption) {
```
Comment (Xenonym) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: A cleaner way of solving this without jamming the sort option into every `Person` is to have this take a `Comparator<Person>`, and use [Comparator.comparing](https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html#comparing-java.util.function.Function-) to create the Comparator for the required sort key in the command.

**Review by Xenonym — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -478,10 +478,41 @@ image::UndoRedoActivityDiagram.png[width="200"]
...
+image::addCommandDemo.png[width="800"]
+
+After: Only name is compulsory for a successful add command +
+
+image::EnhancedAddDemo.png[width="800"]
+
+
 A separate parser for phone, email, address and remark are used to accommodate this enhancement to isolate the interdependency between the modified enhanced add command with the other commands. +
 
+An example is the `parsePhone` method which are also shared with `editCommand`, hence modifying the `parsePhone` method will cause a dependency issue for `editCommand` as well since they requires different parsing method after `addCommand` is enhanced. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: **is**

File: docs/DeveloperGuide.adoc [hunk 5]
```diff
@@ -478,10 +478,41 @@ image::UndoRedoActivityDiagram.png[width="200"]
...
+
+Therefore a separate `parseAddPhone` method is created so that such dependency can be avoided. This is also being implemented to all the other non compulsory fields (email, address, remark).
+
+image::ParsePhone.png[width="800"]
+
+
+It works by instead of returning an empty Optional field, example an empty `Phone` object which sets the internal phone strings to null, we used a substitute string "unspecified Phone" instead. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: What...does this line mean?

File: docs/DeveloperGuide.adoc [hunk 6]
```diff
@@ -478,10 +478,41 @@ image::UndoRedoActivityDiagram.png[width="200"]
...
+
+This implementation ensured that the `Phone` object is not carrying a null string which can cause problems for other components but instead is carrying a predefined "unspecified Phone" string. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: **ensures**

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -478,10 +478,41 @@ image::UndoRedoActivityDiagram.png[width="200"]
...
+
+This is applied across all other optional fields with the exception of name which is made compulsory. +
+
+The initial compulsory checks for prefix is also changed from the original requiring all prefix to be present to the current only requiring name prefix to be compulsory. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: Remove "from the original" and "the current". Change prefix --> prefixes

File: docs/DeveloperGuide.adoc [hunk 8]
```diff
@@ -478,10 +478,41 @@ image::UndoRedoActivityDiagram.png[width="200"]
...
+
+image::CompulsoryPrefixCheck.png[width="800"]
 
 ==== Design Considerations
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc:642]: The ' ' for add command below is wrong. Also, it should be `addCommand`

File: docs/DeveloperGuide.adoc [hunk 9]
```diff
@@ -502,8 +533,46 @@ A separate parser for phone, email, address and remark are used to accommodate t
...
+
+---
+
+**Aspect:** Defining a valid remark input +
+**Alternative 1 (current choice):** No restrictions on the kind of remark input the user can put into the addressbook, i.e. the user can put in any characters in any format. +
+**Pros:** More freedom for the user as the user can type in anything for adding a remark. It is also easier to implement as the input string does not have to go through other checks. +
+**Cons:** Can get messy if the user is not consistent in the way they place in remarks. The sort command applied under remarks also does not sort the remark as well if the input is not consistent. +
+**Alternative 2:** Impose a restriction on remarks in the format [Remark Type] remark followed by restriction of only allowing alphabets. +
+**Pros:** . The sort remark command will be able to organise and sort the remark better since each remarks are predefined at the start in the [Remark Type].+
+**Cons:** . More difficult to implement as we need additional checks to be placed for the restrictions and also less freedom for the user as they need to follow stricter rules before putting in a remark entry.+
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: don't need plus sign at the end of cons

File: docs/DeveloperGuide.adoc [hunk 10]
```diff
@@ -502,8 +533,46 @@ A separate parser for phone, email, address and remark are used to accommodate t
...
 **Cons:** The Tag may be more difficult to find and more work is needed to change the find tag command. Additionally, the user interface of tag command also needs to be changed to accommodate long tags. +
+
+---
+
+**Aspect:** Defining a valid remark input +
+**Alternative 1 (current choice):** No restrictions on the kind of remark input the user can put into the addressbook, i.e. the user can put in any characters in any format. +
+**Pros:** More freedom for the user as the user can type in anything for adding a remark. It is also easier to implement as the input string does not have to go through other checks. +
+**Cons:** Can get messy if the user is not consistent in the way they place in remarks. The sort command applied under remarks also does not sort the remark as well if the input is not consistent. +
+**Alternative 2:** Impose a restriction on remarks in the format [Remark Type] remark followed by restriction of only allowing alphabets. +
+**Pros:** . The sort remark command will be able to organise and sort the remark better since each remarks are predefined at the start in the [Remark Type].+
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: Needs a spacing between the full stop and the plus.

File: docs/DeveloperGuide.adoc [hunk 11]
```diff
@@ -502,8 +533,46 @@ A separate parser for phone, email, address and remark are used to accommodate t
...
+**Cons:** . More difficult to implement as we need additional checks to be placed for the restrictions and also less freedom for the user as they need to follow stricter rules before putting in a remark entry.+
 // end::EnhancedAddCommand[]
 
+//tag::SortCommand[]
+=== Sort Command mechanism
+
+The Sort Command ensures encapsulation as the determining factor of which sort to use is only decided at the layer closest to where the data of the addressbook is. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: Please paraphrase this sentence...we cannot really understand it

File: docs/DeveloperGuide.adoc [hunk 12]
```diff
@@ -502,8 +533,46 @@ A separate parser for phone, email, address and remark are used to accommodate t
...
+
+The `SortCommandParser` will parse the respective PREFIX that the user typed ( n/, p/, e/, a/, r/) and returns an enum of Option corresponding to the various sort types to `SortCommand`. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: `PREFIX`, `n/`, `p/`, `e/`, `a/`, `r/`

File: docs/DeveloperGuide.adoc [hunk 13]
```diff
@@ -502,8 +533,46 @@ A separate parser for phone, email, address and remark are used to accommodate t
...
+
+image::SortCommandParser.png[width="800"]
+
+The passed Option will then be deciphered in `UniquePersonList` which will instruct the `Person` object to activate the respective sort's comparator in each `Person` object. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: "parsed" Option

File: docs/DeveloperGuide.adoc [hunk 14]
```diff
@@ -502,8 +533,46 @@ A separate parser for phone, email, address and remark are used to accommodate t
...
+
+image::SortOptionDecide.png[width="800"]
+
+Each `Person` will then use the passed Option to sort according to the desired comparable. +
```
Comment (joanneong) [inline docs/DeveloperGuide.adoc]: "parsed"

File: src/main/java/seedu/address/logic/commands/SortCommand.java [hunk 15]
```diff
@@ -1,5 +1,10 @@
 package seedu.address.logic.commands;
 //@@author TanYikai
```
Comment (joanneong) [inline src/main/java/seedu/address/logic/commands/SortCommand.java:2]: haha this annotation should be right above the javadoc

**Review by joanneong — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.adoc (hunk 5 above)
Comment (TanYikai) [reply docs/DeveloperGuide.adoc]: The empty optional phone object created sets its own phone string attribute to null

File: docs/DeveloperGuide.adoc (hunk 9 above)
Comment (TanYikai) [reply docs/DeveloperGuide.adoc]: The other fields have a plus sign at the end of the cons. Eg UndoRedo

File: docs/DeveloperGuide.adoc (hunk 14 above)
Comment (TanYikai) [reply docs/DeveloperGuide.adoc]: what I meant was the sortOption is being passed from one class to another after parsing.

File: src/main/java/seedu/address/logic/commands/SortCommand.java (hunk 15 above)
Comment (TanYikai) [reply src/main/java/seedu/address/logic/commands/SortCommand.java:2]: I dont understand what you meant, could you elaborate thanks

File: src/main/java/seedu/address/logic/commands/SortCommand.java [hunk 16]
```diff
@@ -10,9 +13,21 @@
 
     public static final String MESSAGE_SUCCESS = "Sorted all persons";
 
+    private int sortOption;
```
[the remark this answers is not in the document]
Comment (TanYikai) [reply src/main/java/seedu/address/logic/commands/SortCommand.java]: Alright thx for the suggestion

**Pull request merged**